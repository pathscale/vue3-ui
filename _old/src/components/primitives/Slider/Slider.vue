<script>
import { computed, ref, watchEffect } from 'vue'

export default {
  name: 'VSlider',
  props: {
    min: {
      type: [Number, String],
      default: 0,
    },
    max: {
      type: [Number, String],
      default: 100,
    },
    step: {
      type: [Number, String],
      default: 1,
    },
    type: {
      type: String,
    },
    size: String,
    tooltip: {
      type: Boolean,
    },
    rounded: {
      type: Boolean,
    },
    disabled: {
      type: Boolean,
    },
    vertical: Boolean,
    modelValue: [String, Number],
  },
  emits: ['update:modelValue'],
  setup(props, { emit }) {
    const input = ref(null)
    const orient = computed(() => (props.vertical ? 'vertical' : 'horizontal'))
    const position = ref(0)

    const value = ref(props.modelValue)

    watchEffect(() => {
      value.value = props.modelValue
    })

    watchEffect(() => {
      emit('update:modelValue', value.value)
    })

    watchEffect(() => {
      if (input.value) {
        position.value =
          ((value.value - props.min) * input.value.clientWidth) / (props.max - props.min) - 20
      }
    })
    return { value, orient, input, position }
  },
}
</script>

<template>
  <div>
    <input
      ref="input"
      class="slider is-fullwidth"
      :class="[
        type,
        size,
        {
          'is-circle': rounded,
          'has-output-tooltip': tooltip,
        },
      ]"
      :step="step"
      :min="min"
      :max="max"
      v-model="value"
      type="range"
      :disabled="disabled"
      :orient="orient" />
    <output v-if="tooltip" for="sliderWithValueTooltip" :style="`left: ${position}px`">{{
      value
    }}</output>
  </div>
</template>
