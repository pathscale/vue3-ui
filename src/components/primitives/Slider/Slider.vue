<script setup lang="ts">
import { computed, ref, useTemplateRef, watchEffect } from "vue";

const props = withDefaults(
  defineProps<{
    min?: number | string;
    max?: number | string;
    step?: number | string;
    type?: "is-info" | "is-success" | "is-warning" | "is-danger";
    size?: "is-small" | "is-medium" | "is-large";
    tooltip?: boolean;
    rounded?: boolean;
    disabled?: boolean;
    vertical?: boolean;
  }>(),
  {
    min: 0,
    max: 100,
    step: 1,
  },
);

const value = defineModel<number | string>();

const input = useTemplateRef<HTMLInputElement>("input");

/**
 * Non-standard attribute
 * https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/input/range#orient
 */
const orient = computed(() => (props.vertical ? "vertical" : "horizontal"));

const position = ref(0);

watchEffect(() => {
  if (input.value) {
    position.value =
      ((value.value - props.min) * input.value.clientWidth) /
        (props.max - props.min) -
      20;
  }
});
</script>

<template>
  <div>
    <input
      ref="input"
      class="slider is-fullwidth"
      :class="[
        type,
        size,
        {
          'is-circle': rounded,
          'has-output-tooltip': tooltip,
        },
      ]"
      :step="step"
      :min="min"
      :max="max"
      v-model="value"
      type="range"
      :disabled="disabled"
      v-bind="{ orient }" />
    <output v-if="tooltip" for="sliderWithValueTooltip" :style="`left: ${position}px`">{{
      value
    }}</output>
  </div>
</template>
