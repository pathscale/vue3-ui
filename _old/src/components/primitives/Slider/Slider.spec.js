import { shallowMount } from '@vue/test-utils'
import Slider from './Slider.vue'

let wrapper

describe('Slider', () => {
  beforeEach(() => {
    wrapper = shallowMount(Slider)
  })

  it('is called', () => {
    expect(wrapper.exists()).toBeTruthy()
  })

  it('render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
