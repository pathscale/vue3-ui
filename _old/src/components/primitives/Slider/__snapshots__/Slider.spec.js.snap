// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider render correctly 1`] = `
<div><input class="slider is-fullwidth" step="1" min="0" max="100" type="range" orient="horizontal">
  <!--v-if-->
</div>
`;
